dpcpp/factorization/par_ilut_spgeam_kernel.dp.cpp

```diff
@@ -0,0 +1,468 @@
+namespace kernels {
+namespace dpcpp {
+/**
+ * @brief The parallel ILUT factorization namespace.
+ *
+ * @ingroup factor
+ */
+namespace par_ilut_factorization {
+
+
+constexpr int default_block_size = 256;
+
+
+// subwarp sizes for add_candidates kernels
+using compiled_kernels = syn::value_list<int, 1, 8, 16, 32>;
+
+
+namespace kernel {
+
+
+template <int subwarp_size, typename IndexType>
+void tri_spgeam_nnz(const IndexType* __restrict__ lu_row_ptrs,
+                    const IndexType* __restrict__ lu_col_idxs,
+                    const IndexType* __restrict__ a_row_ptrs,
+                    const IndexType* __restrict__ a_col_idxs,
+                    IndexType* __restrict__ l_new_row_ptrs,
+                    IndexType* __restrict__ u_new_row_ptrs, IndexType num_rows,
+                    sycl::nd_item<3> item_ct1)
+{
+    auto subwarp = group::tiled_partition<subwarp_size>(
+        group::this_thread_block(item_ct1));
+    auto row = thread::get_subwarp_id_flat<subwarp_size, IndexType>(item_ct1);
+    if (row >= num_rows) {
+        return;
+    }
+
+    auto lu_begin = lu_row_ptrs[row];
+    auto lu_size = lu_row_ptrs[row + 1] - lu_begin;
+    auto a_begin = a_row_ptrs[row];
+    auto a_size = a_row_ptrs[row + 1] - a_begin;
+    IndexType l_count{};
+    IndexType u_count{};
+    group_merge<subwarp_size>(
+        a_col_idxs + a_begin, a_size, lu_col_idxs + lu_begin, lu_size, subwarp,
+        [&](IndexType a_nz, IndexType a_col, IndexType lu_nz, IndexType lu_col,
+            IndexType out_nz, bool valid) {
+            auto col = min(a_col, lu_col);
+            // count the number of unique elements being merged
+            l_count +=
+                popcnt(subwarp.ballot(col <= row && a_col != lu_col && valid));
+            u_count +=
+                popcnt(subwarp.ballot(col >= row && a_col != lu_col && valid));
+            return true;
+        });
+    if (subwarp.thread_rank() == 0) {
+        l_new_row_ptrs[row] = l_count;
+        u_new_row_ptrs[row] = u_count;
+    }
+}
+
+template <int subwarp_size, typename IndexType>
+void tri_spgeam_nnz(dim3 grid, dim3 block, size_type dynamic_shared_memory,
+                    sycl::queue* queue, const IndexType* lu_row_ptrs,
+                    const IndexType* lu_col_idxs, const IndexType* a_row_ptrs,
+                    const IndexType* a_col_idxs, IndexType* l_new_row_ptrs,
+                    IndexType* u_new_row_ptrs, IndexType num_rows)
+{
+    queue->parallel_for(
+        sycl_nd_range(grid, block), [=
+    ](sycl::nd_item<3> item_ct1) [[sycl::reqd_sub_group_size(subwarp_size)]] {
+            tri_spgeam_nnz<subwarp_size>(lu_row_ptrs, lu_col_idxs, a_row_ptrs,
+                                         a_col_idxs, l_new_row_ptrs,
+                                         u_new_row_ptrs, num_rows, item_ct1);
+        });
+}
+
+
+template <int subwarp_size, typename ValueType, typename IndexType>
+void tri_spgeam_init(const IndexType* __restrict__ lu_row_ptrs,
```

Comment:
Honestly, it's very much possible that you can get similar/better performance from a thread-per-row parallelization of this algorithm. Might be interesting to investigate this at some point, This implementation was very fun do write, but may also be overkill.